provers/risc0/driver/src/bonsai/auto_scaling.rs

```diff
@@ -0,0 +1,166 @@
+use anyhow::{Error, Result};
+use lazy_static::lazy_static;
+use reqwest::{header::HeaderMap, header::HeaderValue, header::CONTENT_TYPE, Client};
+use serde::Deserialize;
+use std::env;
+use tracing::{debug, error as trace_err};
+
+#[derive(Debug, Deserialize, Default)]
+struct ScalerResponse {
+    desired: u32,
+    current: u32,
+    pending: u32,
+}
+struct BonsaiAutoScaler {
+    url: String,
+    headers: HeaderMap,
+}
+
+impl BonsaiAutoScaler {
+    fn new(bonsai_api_url: String, api_key: String) -> Self {
+        let url = bonsai_api_url + "/workers";
+        let mut headers = HeaderMap::new();
+        headers.insert(CONTENT_TYPE, HeaderValue::from_static("application/json"));
+        headers.insert("x-api-key", HeaderValue::from_str(&api_key).unwrap());
+        Self { url, headers }
+    }
+
+    async fn get_bonsai_gpu_num(&self) -> Result<ScalerResponse> {
+        // Create a new client
+        let client = Client::new();
+
+        debug!("Requesting scaler status from: {}", self.url);
+        // Make the POST request
+        let response = client
+            .get(self.url.clone())
+            .headers(self.headers.clone())
+            .send()
+            .await?;
+
+        // Check if the request was successful
+        if response.status().is_success() {
+            // Parse the JSON response
+            let data: ScalerResponse = response.json().await.unwrap_or_default();
+            debug!("Scaler status: {:?}", data);
+            Ok(data)
+        } else {
+            trace_err!("Request failed with status: {}", response.status());
+            Err(Error::msg("Failed to get bonsai gpu num".to_string()))
+        }
+    }
+
+    async fn set_bonsai_gpu_num(&self, gpu_num: u32) -> Result<()> {
+        // Create a new client
+        let client = Client::new();
+
+        // Make the POST request
+        let response = client
+            .post(self.url.clone())
+            .headers(self.headers.clone())
+            .body(gpu_num.to_string())
+            .send()
+            .await?;
+
+        // Check if the request was successful
+        if response.status().is_success() {
+            // Parse the JSON response
+            let data: ScalerResponse = response.json().await?;
+
+            // Use the parsed data
+            debug!("Scaler status: {:?}", data);
+            assert_eq!(data.desired, gpu_num);
+            Ok(())
+        } else {
+            trace_err!("Request failed with status: {}", response.status());
+            Err(Error::msg("Failed to get bonsai gpu num".to_string()))
+        }
+    }
+}
+
+lazy_static! {
+    static ref BONSAI_API_URL: String =
+        env::var("BONSAI_API_URL").expect("BONSAI_API_URL must be set");
+    static ref BONSAI_API_KEY: String =
+        env::var("BONSAI_API_KEY").expect("BONSAI_API_KEY must be set");
+}
+
+const MAX_BONSAI_GPU_NUM: u32 = 15;
```

Comment:
This should be a `env` variable so we don't need to recompile if there is a change in available `gpu`s

Reply:
sounds good, will change.

Reply:
added.